Add OVHcloud Getting Started (#26)
sighupio/fury-getting-started: fury-on-ovhcloud/README.md

2. Download the latest version of Fury with `furyctl`
3. Install the Fury distribution
4. Explore some features of the distribution
5. Teardown of the environment

> ⚠️ OVHcloud **charges you** to provision the resources used in this tutorial. You should be charged only a few euros, but we are not responsible for any costs that incur.
>
> ❗️ **Remember to remove all the instances by following all the steps listed in the teardown phase.**
>
> 💻 If you prefer trying Fury in a local environment, check out the [Fury on Minikube][fury-on-minikube] tutorial.

## Prerequisites

This tutorial assumes some basic familiarity with Kubernetes and [OVHcloud](https://www.ovhcloud.com/). Some experience with Terraform is helpful but not required.

To follow this tutorial, you need:

- **OVHcloud Account** - You must have an active account to use OVHcloud services. Use this [guide](https://docs.ovh.com/gb/en/customer/create-ovhcloud-account/) to create one.
- **OVHcloud Public Cloud project** - You must have a Public Cloud Project. Use this [guide](https://docs.ovh.com/gb/en/public-cloud/create_a_public_cloud_project/) to create one.

> - **OVHcloud Openstack User** - To manage the network environment with the Terraform provider, you must have an Openstack user. Use this [guide](https://docs.ovh.com/gb/en/public-cloud/creation-and-deletion-of-openstack-user/) to create one.

Author: Done (thx)

## Step 1 - Automatic provisioning of a Managed Kubernetes Cluster in a private network

> We are using the `terraform` cli to automaticly deploy the private network, and then the Managed Kubernetes Cluster.

Author: Done (thx)

| Terraform Provider | Credentials |
|---|---|
| [OVH Provider](https://registry.terraform.io/providers/ovh/ovh/latest/docs) | [OVHcloud API](https://docs.ovh.com/gb/en/api/first-steps-with-ovh-api/) credentials |

> | [Openstack Provider](https://registry.terraform.io/providers/terraform-provider-openstack/openstack/latest/docs) | [Openstack user](https://docs.ovh.com/sg/en/public-cloud/creation-and-deletion-of-openstack-user/) credentials |

Author: I don't see the meaning of this one :)

### Pre-requisites

The tools we need are `furyctl`, `terraform`, `kubectl` and `kustomize`.

> 

Reviewer: We should mention that you can also use the container that we make, like in the other guides.

And that there's a script provided to run all the installation commands at once.

Click on the desired tool to see how to install it:

<details><summary>1 - furyctl</summary>

> Install the latest `furyctl` cli from the [Github Furyctl Realise page](https://github.com/sighupio/furyctl/releases).

Author: Done (thx)

Example on an Ubuntu distribution:

```bash
wget -q "https://github.com/sighupio/furyctl/releases/download/v0.9.0/furyctl-$(uname -s)-amd64" -O /tmp/furyctl \
&& chmod +x /tmp/furyctl \
&& sudo mv /tmp/furyctl /usr/local/bin/furyctl
```

> 

Reviewer: ```suggestion

See [furyctl's readme](https://github.com/sighupio/furyctl) for more installation methods.

```

</details>

> <details><summary>2 - terraform cli</summary>

Author: Done (thx)

export OVH_APPLICATION_SECRET="xxxxxxxxxxxxxxxxx"
export OVH_CONSUMER_KEY="xxxxxxxxxxxxxxxxxxxxxxx"
```

You also need to define the API endpoint and base url. Choose the values inside this table according to your localization:

| OVH_ENDPOINT | OVH_BASEURL |
|---|---|
| ovh-eu | https://eu.api.ovh.com/1.0 |
| ovh-us | https://api.us.ovhcloud.com/1.0 |
| ovh-ca | https://ca.api.ovh.com/1.0 |
| kimsufi-eu | https://eu.api.kimsufi.com/1.0 |
| kimsufi-ca | https://ca.api.kimsufi.com/1.0 |
| soyoustart-eu | https://eu.api.soyoustart.com/1.0 |
| soyoustart-ca | https://ca.api.soyoustart.com/1.0 |

Example for `eu` zone:

```bash
export OVH_ENDPOINT=ovh-eu
export OVH_BASEURL="https://eu.api.ovh.com/1.0"
```

The last variable needed by the provider is your Public cloud Id. You get it directly from your [OVHcloud Public Cloud dashboard](https://www.ovh.com/manager/#/public-cloud/pci/projects/xxxxxxxxxxxxxxxxxxxxxxx), the `xxxxxxxxxxxxxx` pert of the url or on the dashboard just under the project name.

Example:

```bash
export OVH_CLOUD_PROJECT_SERVICE="xxxxxxxxxxxxxxxxxxxxx"
```

That's all you need to use the OVHcloud Terraform Provider.

> #### 2 - OVHcloud Openstack User variables

Author: Done (thx)

> [Get your Openstack user's `openrc` file](https://docs.ovh.com/gb/en/public-cloud/set-openstack-environment-variables/#step-1-retrieve-the-variables) to extract and to set necessary variables:

Author: Done (thx)

```bash
export OS_AUTH_URL=https://auth.cloud.ovh.net/v3
export OS_IDENTITY_API_VERSION=3
export OS_USER_DOMAIN_NAME=${OS_USER_DOMAIN_NAME:-"Default"}
export OS_PROJECT_DOMAIN_NAME=${OS_PROJECT_DOMAIN_NAME:-"Default"}
export OS_TENANT_ID="xxxxxxxxxxxxxxxxxx"
export OS_TENANT_NAME="xxxxxxxxxxxxxxxx"
export OS_USERNAME="user-xxxxxxxxxxxxxx"
export OS_PASSWORD="xxxxxxxxxxxxxxxxxxx"
export OS_REGION_NAME="xxx"
```

> You are ready to use the Openstack Terraform Provider.

Author: Done (thx)

# Network - Private Network

network = {
  name = "furyNetwork"
}

# Network - Subnet

subnet = {
  name       = "furySubnet"
  cidr       = "10.0.0.0/24"
  dhcp_start = "10.0.0.100"
  dhcp_end   = "10.0.0.254"
}

# Network - Router

router = {
  name = "furyRouter"
}

# Managed Kubernetes Cluster

kube = {
  name            = "furykubernetesCluster"
  pv_network_name = "furyNetwork"
  gateway_ip      = "10.0.0.1"
}

pool = {
  name          = "furypool"
  flavor        = "b2-7"
  desired_nodes = "3"
  max_nodes     = "6"
  min_nodes     = "3"
}
```

- `region`: The main region where you want to deploy your infrastructure.

- `network`: The network name.

- `subnet`: The subnet parameters, like the CIDR IP format value and DHCP range.

- `router`: The router name.

- `kube`: The Managed Kubernetes Cluster, such as its name and essentially network informations.

- `pool`: The Kubernetes node pool parameters.

Deploy the infrastructure with:

```bash
cd infrastructure/terraform

terraform init
terraform plan -var-file=variables.tfvars
terraform apply -var-file=variables.tfvars
```

Wait a few minutes until the end of the deployement.

### Configure your kubectl environment

> Once the Managed Kubernetes Cluster created, in the `terraform` folder, you get the associated `kubeconfig` file. Set the KUBECONFIG variable value like this:

Author: Done (thx)

```bash

> export KUBECONFIG="$(pwd)/kubeconfig"

Reviewer: better to use $PWD (an env var), than catching the output of a subshell with `$()`:
```suggestion
export KUBECONFIG="$PWD/kubeconfig"
```

Author: Done (thx)

  - ../../vendor/katalog/logging/opensearch-dashboards

  - resources/ingress.yml

patchesStrategicMerge:
  - patches/opensearch-resources.yml
  - patches/cerebro-resources.yml
```

Each `kustomization.yaml`:

- references the modules downloaded in the previous section
- patches the upstream modules (e.g. `patches/opensearch-resources.yml` limits the resources requested by OpenSearch)
- deploys some additional custom resources (e.g. `resources/ingress.yml`)

Install the modules:

```bash
cd manifests/

make apply
# Due to some chicken-egg 🐓🥚 problem with custom resources you have to apply multiple times
make apply
```

## Step 4 - Explore the distribution

🚀 The distribution is finally deployed! In this section, you explore some of its features.

### Setup local DNS

In Step 3, alongside the distribution, you have deployed Kubernetes ingresses to expose underlying services at the following HTTP routes:

- `forecastle.fury.info`
- `grafana.fury.info`
- `opensearch-dashboards.fury.info`

To access the ingresses more easily via the browser, configure your local DNS to resolve the ingresses to the external load balancer IP:

1. Get the address of the external load balancer:

```bash

> kubectl get svc -n ingress-nginx ingress-nginx --no-headers | awk '{print $4}'

Reviewer: I NEED TO TEST THIS, but I'm 99% sure it works

No need to pipe to `awk`, you can tell kubectl what to print out using the `-o jsonpath=<template>` flag:
```suggestion
kubectl get svc -n ingress-nginx ingress-nginx -ojsonpath='{.spec.externalIPs[*]}'
```

#### Read the logs

The Fury Logging module already collects data from the following indices:

- `kubernetes-*`
- `systemd-*`
- `ingress-controller-*`
- `events-*`

Click on `Discover` to see the main dashboard. On the top left corner select one of the indices to explore the logs.

![Opensearch-Dashboards][opensearch-dashboards-screenshot]

### Grafana

[Grafana](https://github.com/grafana/grafana) is an open-source platform for monitoring and observability. Grafana allows you to query, visualize, alert on and understand your metrics.

Navigate to <http://grafana.fury.info> or click the Grafana icon from Forecastle.

Fury provides some pre-configured dashboards to visualize the state of the cluster. Examine an example dashboard:

1. Click on the search icon on the left sidebar.
2. Write `pods` and click enter.
3. Select the `Kubernetes/Pods` dashboard.

This is what you should see:

![Grafana][grafana-screenshot]

## Step 5 - Teardown

Clean up the demo environment:

```bash
cd ./terraform/
terraform destroy -var-file=variables.tfvars
```

## Conclusions

Congratulations, you made it! 🥳🥳

> We hope you enjoyed this tour of Fury!

Reviewer: ```suggestion
We hope you enjoyed this tour of Fury on OVHcloud!
```

```yaml
---
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization

resources:
  - ingress
  - logging
  - monitoring
```

This `kustomization.yaml` wraps the other `kustomization.yaml`s in subfolders. For example in `manifests/logging/kustomization.yaml`

```yaml
---
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization

resources:
  - ../../vendor/katalog/logging/cerebro
  - ../../vendor/katalog/logging/logging-operator
  - ../../vendor/katalog/logging/logging-operated
  - ../../vendor/katalog/logging/configs
  - ../../vendor/katalog/logging/opensearch-single
  - ../../vendor/katalog/logging/opensearch-dashboards

  - resources/ingress.yml

patchesStrategicMerge:
  - patches/opensearch-resources.yml
  - patches/cerebro-resources.yml
```

Each `kustomization.yaml`:

- references the modules downloaded in the previous section
- patches the upstream modules (e.g. `patches/opensearch-resources.yml` limits the resources requested by OpenSearch)
- deploys some additional custom resources (e.g. `resources/ingress.yml`)

Install the modules:

```bash
cd manifests/

make apply

> # Due to some chicken-egg 🐓🥚 problem with custom resources you have to apply multiple times

Reviewer: ```suggestion
# Due to some CRDs being created, the first time you have to run make apply multiple times. Run it until you see no more errors.
```
